interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,248 @@
+    DataOutputStream clientOut = new DataOutputStream(server.getOutputStream());
+    DataInputStream serverIn = new DataInputStream(server.getInputStream());
+    DataOutputStream serverOut = new DataOutputStream(clientSock.getOutputStream());
+
+    (new Thread(new ClientReader(clientIn))).start();
+    (new Thread(new ClientWriter(clientOut))).start();
+
+    (new Thread(new ServerReader(serverIn))).start();
+    (new Thread(new ServerWriter(serverOut))).start();
+
+  }
+
+  public void shutDown() {
+    try {
+      System.out.println("Proxy shutting down... ");
+      server.close();
+      client.close();
+    } catch (IOException e) {
+      System.out.println("Shutdown failed");
+      e.printStackTrace();
+    }
+  }
+
+  private synchronized void incrementServerQSize(int amt) {
+    serverQSize += amt;
+  }
+
+  private synchronized void incrementClientQSize(int amt) {
+    clientQSize += amt;
+  }
+
+  public static void main(String[] args) {
+    TrafficControlProxy p = new TrafficControlProxy();
+    try {
+      p.setUp();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  private class ClientReader implements Runnable {
+
+    DataInputStream clientIn;
+
+    public ClientReader(DataInputStream clientIn) {
+      this.clientIn = clientIn;
+    }
+
+    @Override
+    public void run() {
+      while (true) {
+        try {
+          byte[] request = new byte[chunkSize];
+          int readableRequestBytes = clientIn.read(request);
+          long arrivalTime = System.currentTimeMillis();
+          Message m = new Message(arrivalTime, request, readableRequestBytes);
+          while (clientQSize + readableRequestBytes > qLength) {
+            System.out.println(clientQSize + readableRequestBytes);
+            Thread.sleep(1);
+          }
+          clientQueue.add(m);
+          incrementClientQSize(readableRequestBytes);
+        } catch (IOException e) {
+          e.printStackTrace();
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+      }
+
+    }
+
+  }
+
+  private class ServerReader implements Runnable {
+
+    DataInputStream serverIn;
+
+    public ServerReader(DataInputStream serverIn) {
+      this.serverIn = serverIn;
+    }
+
+    @Override
+    public void run() {
+      while (true) {
+        try {
+          byte[] response = new byte[chunkSize];
+          int readableResponseBytes = serverIn.read(response);
+          long arrivalTime = System.currentTimeMillis();
+          Message m = new Message(arrivalTime, response, readableResponseBytes);
+          while (serverQSize + readableResponseBytes > qLength) {
+            Thread.sleep(1);
+          }
+          System.out.println("Queueing a server message");
+          serverQueue.add(m);
+          incrementServerQSize(readableResponseBytes);
+          System.out.println("Server q size: " + serverQSize);
+        } catch (IOException e) {
+          e.printStackTrace();
```

Comment:
Just let these exceptions propagate and set an Uncaught exception handler on the thread.